core/templates/pages/exploration-player-page/services/diagnostic-test-player-engine.service.ts

```diff
@@ -0,0 +1,419 @@
+          tempStack = nonVisitedPrerequisites.concat(tempStack);
+        } else {
+          visitedTopicIds.push(topicId);
+        }
+      }
+    }
+    return visitedTopicIds;
+  }
+
+  computeProgressPercentage(): number {
+    let numberOfAttemptedQuestionsInCurrentTopic = (
+      this._diagnosticTestCurrentTopicStatusModel.numberOfAttemptedQuestions);
+    let initialTopicIdsList = (
+      this._initialCopyOfTopicTrackerModel.getEligibleTopicIds());
+    let pendingTopicIdsToTest = (
+      this._diagnosticTestTopicTrackerModel.getEligibleTopicIds());
+
+    // Each topic can contain a maximum of 3 diagnostic test skills and at most
+    // 2 questions [main question & backup question] can be presented from each
+    // skill. Thus, the maximum number of questions that can be asked from a
+    // topic is 6.
+    let completionMetric = (((
+      initialTopicIdsList.length - pendingTopicIdsToTest.length) * 6 +
+        numberOfAttemptedQuestionsInCurrentTopic) / (
+      initialTopicIdsList.length * 6));
+
+    return Math.round(completionMetric * 100);
+  }
+
+  getLanguageCode(): string {
+    return (
+      this._diagnosticTestCurrentTopicStatusModel.getNextQuestion(
+        this._currentSkillId).getLanguageCode()
+    );
+  }
+
+  recordNewCardAdded(): void {
+    this.contextService.setCustomEntityContext(
+      AppConstants.ENTITY_TYPE.QUESTION, this._currentQuestion.getId() as string
+    );
+  }
+
+  createCard(question: Question): StateCard {
+    this._encounteredQuestionIds.push(question.getId() as string);
+
+    const stateData: State = question.getStateData();
+    const questionHtml: string = (
+      this.expressionInterpolationService.processHtml(
+        stateData.content.html, [])
+    );
+
+    if (questionHtml === '') {
+      this.alertsService.addWarning('Question name should not be empty.');
+    }
+
+    this._focusLabel = this.focusManagerService.generateFocusLabel();
+    const interaction = stateData.interaction;
+    const interactionId = interaction.id;
+    let interactionHtml: string = '';
+
+    if (interactionId) {
+      interactionHtml = (
+        this.explorationHtmlFormatterService.getInteractionHtml(
+          interactionId, interaction.customizationArgs,
+          true, this._focusLabel, null)
+      );
+    }
+
+    return StateCard.createNewCard(
+      stateData.name as string, questionHtml, interactionHtml as string,
+      interaction, stateData.recordedVoiceovers, stateData.writtenTranslations,
+      stateData.content.contentId as string,
+      this.audioTranslationLanguageService
+    );
+  }
+
+  static get MAX_ALLOWED_QUESTIONS_IN_THE_DIAGNOSTIC_TEST(): number {
+    return AppConstants.MAX_ALLOWED_QUESTIONS_IN_THE_DIAGNOSTIC_TEST;
+  }
+
+  isDiagnosticTestFinished(): boolean {
+    const pendingTopicIdsToTest = (
+      this._diagnosticTestTopicTrackerModel.getEligibleTopicIds().length
+    );
+
+    if (pendingTopicIdsToTest === 0) {
+      return true;
+    }
+    if (
```

Comment:
At this point, pendingTopicIdsToTest must be > 0, so the logic would be clearer if it was:

```
    if (pendingTopicIdsToTest === 0) {
      return true;
    } else if (this._numberOfAttemptedQuestions >= DiagnosticTestPlayerEngineService
          .MAX_ALLOWED_QUESTIONS_IN_THE_DIAGNOSTIC_TEST) {
      return true;
    } else {
      return false;
    }
```

Reply:
Thank you, updated